### PR TITLE
Issue #2: Review sink EnturSales

### PR DESCRIPTION
**Pull request by jbtfk**
New sink, needs review
File list:

- config/ragnarok_entur.php
- database/migrations/2024_03_14_100000_entur_product_dales.php
- src/Facades/EnturCleosApi.php
- src/Services/EnturSales.php
- src/Sinks/EnturSales.php
- README.md

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -38,10 +38,20 @@ Makes Entur data available, so far only NeTEx route data.
    ENTUR_API_PATH          = timetable-admin/v1/timetable/download_netex_blocks/***
    ENTUR_ENV               = staging
 
+   CLEOS_ENV               = staging
+   CLEOS_AUTH_URL          = https://partner.staging.entur.org/oauth/token
+   CLEOS_CLIENT_ID         = ***
+   CLEOS_SECRET            = ***
+   CLEOS_AUDIENCE          = https://api.staging.entur.io
+   CLEOS_API_PATH          = cleos-reporting/api/v1
+
```
Comment (tfk-kaare) [inline README.md]: Is the cleos account different from our ENTUR_API account? Can we use the same account?

If not, all `CLEOS_*` vars should be prefixed with `ENTUR_`

File: README.md [hunk 2]
```diff
@@ -38,10 +38,20 @@ Makes Entur data available, so far only NeTEx route data.
...
    # Main XML file for parsing netex route data.
    ENTUR_ROUTEDATA_XML_MAIN = ***shared_data.xml
+
+
```
Comment (tfk-kaare) [inline README.md:50]: unnecessary whitespace

File: src/Services/CleosAuthToken.php
Comment (tfk-kaare) [inline src/Services/CleosAuthToken.php:1]: It would help tremendous if we could use a combined Entur account for route data and cleos. Then there will be no need for this file.

Have you/Truls talked to Entur about this?

File: src/Services/EnturSales.php [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        $mapper->column('LINE_CATEGORY_DESCRIPTION', 'line_category_description');
+        $mapper->column('LINE_AMOUNT', 'line_amount');
+        $mapper->column('LINE_CANCELLATION', 'line_cancellation');
+        $mapper->column('LINE_STANDARD_TAX_CODE', 'line_standard_tax_code');
+        $mapper->column('LINE_LOCAL_TAX_CODE', 'line_local_tax_code');
+        $mapper->column('LINE_LOCAL_TAX_RATE', 'line_local_tax_rate');
+        $mapper->column('LINE_TAX_AMOUNT', 'line_tax_amount');
+
+
+
```
Comment (tfk-kaare) [inline src/Services/EnturSales.php:209]: Too much white space

File: src/Sinks/SinkEnturSales.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+use Ragnarok\Entur\Services\EnturSales;
+use Ragnarok\Sink\Models\SinkFile;
+
+class SinkEnturSales extends SinkEnturBase
+{
+    public static $id = "entur-sales";
+    public static $title = "Entur Salesdata";
+    public $cron = '30 04 * * *';
+    //public $cron = '0 21 12 * ?';
+
```
Comment (tfk-kaare) [inline src/Sinks/SinkEnturSales.php]: Remove comment

File: tests/Unit/EnturSalesConstructionTest.php
Comment (tfk-kaare) [inline tests/Unit/EnturSalesConstructionTest.php:1]: Our first test file! Awesome!

File: database/migrations/2024_03_14_100000_entur_product_sales.php
Comment (tfk-kaare) [inline database/migrations/2024_03_14_100000_entur_product_sales.php:1]: All other sinks prefix migrations with `ragnarok_<sink>_<what>`, my suggestion is to name it `..._ragnarok_entur_sales_tables`

File: config/ragnarok_entur.php [hunk 5]
```diff
@@ -18,6 +18,24 @@
...
+    ],
+
+    /*'cleos' => [
+        'auth_url' => env('ENTUR_AUTH_URL'),
+        'client_id' => env('ENTUR_CLIENT_ID'),
+        'client_secret' => env('ENTUR_SECRET'),
+        'audience' => env('ENTUR_AUDIENCE'),
+        'api_path' => trim(env('ENTUR_API_PATH'), '/'),
+        'env' => env('ENTUR_ENV', 'dev'),
+    ],
```
Comment (tfk-kaare) [inline config/ragnarok_entur.php]: Remove debug comments

**Review by tfk-kaare — COMMENTED** [review]
Overall fixes:
- The biggest obstacle is the use of two accounts against Entur. This needs to be absolutely asserted that we need two accounts.
- Second, from a functional view, this sink overrides SinkEnturBase, which are targeted single-state sinks, having lots of custom code. It might be feasible to inherit this sink directly from `SinkBase` and implement `getFromDate()`.
- destinationTables() returns a *keyed* list of tables, where the table name is used for key and the value is the table description.
- Otherwise it's just nitpicks, and a bug in the repository: There should be a commit block on faulty linted code, so staging changes and running `npx lint-staged` will give you the same result. There are a lot of minor lint stuff that need to be fixed. I'll do a quick investigation why this fails.
- To fix lint-staged, run `chmod 775 .husky/pre-commit`

File: src/Sinks/SinkEnturSales.php
Comment (tfk-kaare) [inline src/Sinks/SinkEnturSales.php:1]: The two other Entur sinks are single state sinks, where you can only download and have one valid chunk at the time. This sink has a lot of available chunks and need to have its `$singleState` flag set to false. Also, this sink is missing start date for first available chunk. Trus/Jan knows what this is, but I assume it's around 2024-01-02.

The sink api call `getChunkIds()` is used to return all available chunk IDs from a sink, and as most sinks have chunks split by date, a convenience api `getFromDate()` can be implemented to set this, and voila, all sink IDs are available.

Except, the two sinks `SinkEnturStops` and `SinkEnturRoutes` overrides this in `SinkEnturBase`, and implement its own getChunkIds().

Adding up, you need to move stuff from the `SinkEnturBase`, or have `SinkEnturSales` *not* inherit from `SinkEnturBase` to overcome this.

File: README.md [hunk 6]
```diff
@@ -38,10 +38,20 @@ Makes Entur data available, so far only NeTEx route data.
    ENTUR_API_PATH          = timetable-admin/v1/timetable/download_netex_blocks/***
    ENTUR_ENV               = staging
 
+   CLEOS_ENV               = staging
```
Comment (tfk-kaare) [inline README.md]: Not used

File: README.md (hunk 2 above)
Comment (jbtfk) [reply README.md:50]: fixed

File: src/Services/EnturSales.php (hunk 3 above)
Comment (jbtfk) [reply src/Services/EnturSales.php:209]: Fixed

File: src/Sinks/SinkEnturSales.php (hunk 4 above)
Comment (jbtfk) [reply src/Sinks/SinkEnturSales.php]: fixed

File: config/ragnarok_entur.php (hunk 5 above)
Comment (jbtfk) [reply config/ragnarok_entur.php]: fixed

File: src/Sinks/SinkEnturSales.php
Comment (jbtfk) [reply src/Sinks/SinkEnturSales.php:1]: Fixed, inherits from SinkBase now

File: README.md (hunk 6 above)
Comment (jbtfk) [reply README.md]: Removed CLEOS_ENV from config/ragnarok_entur.php

File: src/Sinks/SinkEnturSales.php [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        // Lets use already provided/stored data as initial date.
+        return Carbon::parse("2024-03-07");
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getToDate(): Carbon
+    {
+        return Carbon::today();
```
Comment (tfk-kaare) [inline src/Sinks/SinkEnturSales.php:73]: Running this just past midnight will probably result in empty data?

We can't access reports for future transactions :-)

**Review by tfk-kaare — COMMENTED** [review]
All good, except getToDate() which should return yesterday, at earliest :-)

**Review by tfk-kaare — APPROVED** [review]
Let's go!
